src/algorithms/clustering/metrics.py

```diff
@@ -0,0 +1,85 @@
+import numpy as np
+import pandas as pd
+
+
+def distance(x, y):
+    return np.sqrt(np.sum(x**2 + y**2))
+
+
+def davies_bouldin_score(df: pd.DataFrame, labels: np.array):
+    num_cluster = np.amax(labels) + 1
+    centroids = []
+    avg_distances = []
+    for cluster in range(num_cluster):
+        cluster_set = df.loc[labels == cluster].to_numpy()
+        centroid = cluster_set.mean(axis=0)
+        avg_distance = np.sqrt(
+            (cluster_set**2 + np.stack([centroid**2] * len(cluster_set))).sum(
+                axis=1
+            )
+        ).mean()
+        centroids.append(centroid)
+        avg_distances.append(avg_distance)
+    R_matrix = np.zeros((num_cluster, num_cluster), dtype=np.float)
+    for i in range(num_cluster):
+        x = centroids[i]
+        for j in range(i + 1, num_cluster):
+            y = centroids[j]
+            R_matrix[i, j] = R_matrix[j, i] = (
+                avg_distances[i] + avg_distances[j]
+            ) / distance(x, y)
+    return np.sum(np.amax(R_matrix, axis=0)) / num_cluster
+
+
+def dunn_score(df: pd.DataFrame, labels: np.array, distance_function=distance):
+    num_cluster = np.amax(labels) + 1
+    centroids = []
+    for cluster in range(num_cluster):
+        cluster_set = df.loc[labels == cluster].to_numpy()
+        centroid = cluster_set.mean(axis=0)
+        centroids.append(centroid)
+    max_distance_intra = 0
+    min_distance_inter = np.inf
+    for i, first in enumerate(centroids):
+        for second in centroids[i + 1 :]:
+            dis = distance_function(first, second)
+            min_distance_inter = min(min_distance_inter, dis)
+    for cluster in range(num_cluster):
+        cluster_df = df.loc[labels == cluster]
+        for i, pointA in cluster_df.iterrows():
+            for _, pointB in cluster_df.loc[i + 1 :].iterrows():
+                dis = distance_function(pointA.to_numpy(), pointB.to_numpy())
+                max_distance_intra = max(max_distance_intra, dis)
+    return min_distance_inter / max_distance_intra
```

Comment:
this line crashes for quite a few non-standard data sets, when `max_distance_intra` is zero. maybe add an if statement to catch this or use a try block around calculating any metric to catch arithmetic errors and set default value or a nan for it instead, cause I think there are more places that can have similar problems

Reply:
Thanks, I forgot about dividing by zero. I will check all the places where I divide by some variables.